### ISSUE TITLE
Bulk export of documents in specific output format

### ISSUE THREAD

**Issue by jfiala** [inception-project/inception#3402]
**Is your feature request related to a problem? Please describe.**
Currently it seems only possible to export one document in the annotation view in a desired format (e.g. Apache UIMA CAS XMI 1.1)

**Describe the solution you'd like**
It would be nice to have the option to export multiple documents (e.g. in Settings - documents list)

**Comment by reckart** [comment]
If you export the entire project from the project settings, all documents are in there.

**Comment by reckart** [comment]
In v25 you will be able to export multiple documents from the documents list in the project settings - but this export will just retrieve the originally uploaded documents again, not the annotated documents.

<img width="1144" alt="Screenshot 2022-09-26 at 15 44 34" src="https://user-images.githubusercontent.com/1410238/192292585-6f5e259c-67ec-46e5-881f-12c255843aa8.png">

I think it might be considered to add export option to the workload management pages...

**Comment by jfiala** [comment]
Thank you for the hint!
However, there are two obstacles here:
_ each document is contained within a zip file with random number file name
e.g.: webanno5173705207688145129export.zip
without a backlink to the document name

_ The filename inside the ZIP-file seems to be hardcoded to "admin.xmi"

This makes the bulk export nearly unusable for XMI-format as there is no backlink to the original document file name?

**Comment by reckart** [comment]
Yeah, I guess the name of that zip file should be fixed to properly represent the document name...

**Comment by reckart** [comment]
Actually, the XMI file itself should contain a "DocumentMetaData" annotation and the name of the document should be in there. Not perfect, but also not a lost cause.

**Comment by jfiala** [comment]
just found it (documentTitle). 
Thats good for a workaround with a postprocessing script, thank you!

**Comment by jfiala** [comment]
I don't know if anybody needs the "webanno...export.zip" file name.
I'd suggest to simply use the document title for the zip and xmi file name (both for document-level and project-level export).

**Comment by jfiala** [comment]
Any plans to add this to the next release or should I write a reformat script meanwhile?

**Comment by jfiala** [comment]
Additionally, for the project-level export I think the TypeSystem.xml would be required only once.
So there could be one zip file with all the documents + one TypeSystem.xml?

**Comment by reckart** [comment]
If you need a solution now, you better write a script. I'll mark the issue for v25, but no promises.

**Comment by reckart** [comment]
Guess this will split at least to v26.

**Comment by jfiala** [comment]
Regarding UIMA JSON: IMHO it is not a good idea to embed the whole type system, as this will blow up the json unnecessarily. 
Especially when you have thousands of documents. The XMI we use currently is a great and compact format.

**Comment by reckart** [comment]
You are absolutely right here and I have noticed that as well. By default, only a minimal type system should be included (i.e. one that covers only the types actually used in the respective JSON file). However, I didn't have the time to investigate yet why the included type system is not minimized and if that is a bug/missing setting in INCEpTION or in the underlying UIMA library.

**Comment by jfiala** [comment]
I think a pointer to the underlying json schema would be best.
That would maybe also allow to keep document properties in a versioned way....

**Comment by reckart** [comment]
Having a pointer would mean the pointer needs to be resolvable...

**Comment by jfiala** [comment]
What about having a pointer to schema + embed the schema in zip file to be always in sync?

**Comment by reckart** [comment]
As said before, having only a single copy of the type system in the export ZIP is currently not possible - it requires refactoring. The current architecture passes each document in turn to the exporter and the exporter is the one that creates the type system - the exporter is not aware that it is being called multiple times and the caller of the exporter has no knowledge that the export might be creating a type system...

Having the type system be large/redundant is a bit annoying but at least it "just" costs space...

**Comment by reckart** [comment]
> Can you create a zip with multiple zips inside as you do for the project export?

In principle yes, but I am not sure if I want to. I'd personally prefer keeping the JSON file as a self-contained file with an embedded type system. My impression is that having a fully self-contained format is also important. It could be considered to offer two export options, one with a shared type system (after some refactoring) and one with redundantly embedded type systems, but at the moment it seems to early to me to do that.

When implementing the bulk-export for XMIs, it will be as you say: ZIP with nested ZIPs, each nested ZIP containing the XMI file and a `typesystem.xml` file - and if you extracted all of these into the same folder, the `typesystem.xml` files would override each other.

**Comment by jfiala** [comment]
Could it be that the type system differs if exported from Inception?

IMHO using a schema with a pointer in each json/xml would be the cleanest solution.

XMI: OK for me, but same issue with same TypeSystem.xml (as from same document storage).
Using an xml schema would have the advantage of validation against the xml files, but is not necessary.

**Comment by reckart** [comment]
> Could it be that the type system differs if exported from Inception?

INCEpTION has the concept of a project-specific type system, so all annotations in a project are expected to adhere to the same type system - though they might be lagging behind because upgrading a document to the project type system is a lazy action that usually happens only when opening the document for annotation.

**Comment by jfiala** [comment]
Thank you for your explanation! I think that lazy propagation of new properties should be mentioned somewhere in the docs (if not already the case)? If an attribut is removed, that is forcefully propagated through all documents.

**Comment by reckart** [comment]
> If an attribut is removed, that is forcefully propagated through all documents.

Yes. 

Otherwise, the lazy update should not make a difference to users because it is an internal behavior. Most internal behavior is implicit in the code and not documented - otherwise there'd be no end to writing documentation.

**Comment by jfiala** [comment]
OK, agreeed thank you !

**Comment by phenningsson** [comment]
Hi, was this issue ever resolved, or is there a relatively quick workaround for exporting multiple annotated documents in, say, CoNLL 2002 format? 

Thanks for allt the great work with INCEpTION!

**Comment by reckart** [comment]
It is possible since 39.x to export documents from the workload management page (static workload). 

* Go to the workload/monitoring page.
* Click on the "bulk action" on the top-right of the table
* Select the rows/columns/cells you want to export the documents for
* Open the dropdown of the bulk action and choose "export"
* Choose the format
* Voilá 

So I guess we can indeed close this issue :)

**Comment by phenningsson** [comment]
Brilliant, thank you ever so much! @reckart
